### PR TITLE
fix bug in wamser sandbox and add wasmer cache

### PR DESCRIPTION
**Pull request by grishasobol**
1) update wasmer version to 2
2) fix bug in wasmer sanbox with state handling: in invoke we used to use state which is defined in instantiate, but state can be changed, so we must use fresh state value during invoke.
3) add wasmer cache

### CODE REVIEW

File: client/executor/common/src/sandbox/wasmer_backend.rs [hunk 1]
```diff
@@ -100,14 +100,60 @@ pub fn invoke(
...
+	let fs_cache =
+		FileSystemCache::new("/tmp/substrate-wasmer-cache").map_err(|_| FileSystemErr)?;
+	let code_hash = Hash::generate(wasm);
+	unsafe { fs_cache.load(store, code_hash).map_err(|_| ModuleLoadErr(fs_cache, code_hash)) }
+}
+
+#[cfg(feature = "wasmer-cache")]
+fn try_to_store_module_in_cache(mut fs_cache: FileSystemCache, code_hash: Hash, module: &Module) {
+	let res = fs_cache.store(code_hash, &module.clone());
+	log::debug!("Store module cache with result: {:?}", res);
```
Comment (ukint-vs) [inline client/executor/common/src/sandbox/wasmer_backend.rs:126]: I think you can remove these log messages.

File: client/executor/common/src/sandbox/wasmer_backend.rs [hunk 2]
```diff
@@ -100,14 +100,60 @@ pub fn invoke(
...
+#[cfg(feature = "wasmer-cache")]
+use CachedModuleErr::*;
+
+#[cfg(feature = "wasmer-cache")]
+fn get_cached_module(
+	wasm: &[u8],
+	store: &wasmer::Store,
+) -> core::result::Result<Module, CachedModuleErr> {
+	let fs_cache =
+		FileSystemCache::new("/tmp/substrate-wasmer-cache").map_err(|_| FileSystemErr)?;
```
Comment (ukint-vs) [inline client/executor/common/src/sandbox/wasmer_backend.rs:118]: maybe use something like `std::env::temp_dir()`? https://doc.rust-lang.org/std/env/fn.temp_dir.html

File: client/executor/common/Cargo.toml [hunk 3]
```diff
@@ -24,13 +24,15 @@ sp-maybe-compressed-blob = { version = "4.1.0-dev", path = "../../../primitives/
 sp-serializer = { version = "4.0.0-dev", path = "../../../primitives/serializer" }
 thiserror = "1.0.30"
 environmental = "1.1.3"
+log = "0.4"
```
Comment (ukint-vs) [inline client/executor/common/Cargo.toml:27]: we shouldn't use logging here.

**Review by ukint-vs — CHANGES_REQUESTED** [review]

File: client/executor/common/src/sandbox/wasmer_backend.rs (hunk 1 above)
Comment (grishasobol) [reply client/executor/common/src/sandbox/wasmer_backend.rs:126]: What?? why?

File: client/executor/common/Cargo.toml (hunk 3 above)
Comment (grishasobol) [reply client/executor/common/Cargo.toml:27]: Why?

File: client/executor/common/src/sandbox/wasmer_backend.rs (hunk 1 above)
Comment (grishasobol) [reply client/executor/common/src/sandbox/wasmer_backend.rs:126]: We may wanna to reuse compiled cache even after node is stopped

File: client/executor/common/src/sandbox/wasmer_backend.rs (hunk 2 above)
Comment (grishasobol) [reply client/executor/common/src/sandbox/wasmer_backend.rs:118]: We may wanna to reuse compiled cache even after node is stopped

Comment (grishasobol) [reply client/executor/common/src/sandbox/wasmer_backend.rs]: Done

File: client/executor/common/src/sandbox/wasmer_backend.rs (hunk 1 above)
Comment (grishasobol) [reply client/executor/common/src/sandbox/wasmer_backend.rs]: Make it as trace

**Review by ukint-vs — APPROVED** [review]

**Pull request merged**